torch/lib/THCUNN/generic/FusedRNNKernel.cu

```diff
@@ -0,0 +1,963 @@
+#ifndef THC_GENERIC_FILE
+#define THC_GENERIC_FILE "generic/FusedRNNKernel.cu"
+#else
+#include <cstdarg>
+
+#include "../common.h"
+
+#define DATATYPE TensorUtils<THCTensor>::DataType
+
+//factor will be 3 for GRU and 4 for LSTM
+void THNN_(FusedRNNAssertSizes)(THCState *state, int factor, int count ...)
+{
+  va_list list;
+  va_start(list, count);
+  THCTensor *input = va_arg(list, THCTensor*);
+  THCTensor *hidden = va_arg(list, THCTensor*);
+  THArgCheck(THCTensor_(nElement)(state, input) ==
+	     THCTensor_(nElement)(state, hidden),
+	     3, "Input and Hidden tensor sizes should be the same.");
+
+  THAssertMsg(TensorUtils<THCTensor>::getDims(state, input) <= MAX_CUTORCH_DIMS,
+		 "Tensor dimension is too large.");
+
+  THAssertMsg(TensorUtils<THCTensor>::getDims(state, hidden) <= MAX_CUTORCH_DIMS,
+		 "Tensor dimension is too large.");
+
+  for (int arg=2; arg < count; ++arg){
+    THCTensor *tens = va_arg(list, THCTensor*);
+    THArgCheck(THCTensor_(nElement)(state, input) ==
+	       THCTensor_(nElement)(state, tens)*factor,
+	       3, "A pointwise tensor was not the right size, should have 1/%u the elements of input/hidden tensor.", arg, factor);
+    THAssertMsg(TensorUtils<THCTensor>::getDims(state, tens) <= MAX_CUTORCH_DIMS,
+		 "Tensor dimension is too large.");
+  }
+
+  va_end(list);
+}
+
+int THNN_(minIndexType)(THCState *state, int count, ...)
+{
+  va_list list;
+  va_start(list, count);
+
+  int maxDim = -2;
+  for (int arg=0; arg < count; ++arg){
+    THCTensor* tens = va_arg(list, THCTensor*);
+    if(THCTensor_(isContiguous)(state, tens)) continue;
+    int tensdims = TensorUtils<THCTensor>::getDims(state, tens);
+    maxDim = (( tensdims> maxDim) ? tensdims : maxDim);
+  }
+
+  va_end(list);
+  return maxDim;
+}
+
+bool THNN_(canUse32BitIndexMath)(THCState *state, int count, ...)
+{
+  va_list list;
+  va_start(list, count);
+
+  for (int arg=0; arg < count; ++arg){
+    THCTensor *tens = va_arg(list, THCTensor*);
+    if (!TensorUtils<THCTensor>::canUse32BitIndexMath(state, tens)){
+	va_end(list);
+	return false;
+      }
+  }
+  va_end(list);
+  return true;
+}
+
+#define DEVICE_LINEAR_GET(D_TENSOR, INDEX)				\
+  D_TENSOR.data[IndexToOffset<T, IndexType, Dims>::get(INDEX, D_TENSOR)]
+
+#define H2F(input) __half2float(input)
+#define F2H(input) __float2half(input)
```

Comment:
@apaszke Would be more accurate to leave everything in floating point and only convert once at beginning and end. Using THCNumerics in half mode would constantly convert back and forth.